src/rust/engine/sharded_lmdb/src/lib.rs

```diff
@@ -378,6 +381,109 @@ impl ShardedLmdb {
       .await
   }
 
+  ///
+  /// Stores the given Read instance under its computed digest. This method performs two passes
+  /// over the source to 1) hash it, 2) store it. If !data_is_immutable, the second pass will
+  /// re-hash the data to confirm that it hasn't changed.
+  ///
+  pub async fn store<F, R>(
+    &self,
+    initial_lease: bool,
+    data_is_immutable: bool,
+    data_provider: F,
+  ) -> Result<Digest, String>
+  where
+    R: Read + Debug,
+    F: Fn() -> Result<R, io::Error> + Send + 'static,
+  {
+    let store = self.clone();
+    self
+      .executor
+      .spawn_blocking(move || {
+        let mut attempts = 0;
+        loop {
+          // First pass: compute the Digest.
+          let digest = {
+            let mut read = data_provider().map_err(|e| format!("Failed to read: {}", e))?;
+            let mut hasher = WriterHasher::new(NoopWriter);
+            let _ = std::io::copy(&mut read, &mut hasher)
+              .map_err(|e| format!("Failed to read from {:?}: {}", read, e))?;
+            hasher.finish().0
+          };
+
+          let effective_key = VersionedFingerprint::new(digest.hash, ShardedLmdb::SCHEMA_VERSION);
+          let (env, db, lease_database) = store.get(&digest.hash);
+          let put_res: Result<(), StoreError> = env
+            .begin_rw_txn()
+            .map_err(StoreError::Lmdb)
+            .and_then(|mut txn| {
+              // Second pass: copy into the reserved memory.
+              let dest = txn.reserve(
+                db,
+                &effective_key,
+                digest.size_bytes,
+                WriteFlags::NO_OVERWRITE,
+              )?;
+              let mut read = data_provider().map_err(|e| format!("Failed to read: {}", e))?;
+              if data_is_immutable {
+                // Trust that the data hasn't changed.
+                read
+                  .read_exact(dest)
+                  .map_err(|e| format!("Failed to read from {:?}: {:?}", read, e))?
+              } else {
+                // Confirm that the data hasn't changed.
```

Comment:
Do we care if the file had been appended to between the hash and the read? `read_exact` won't error if there are _more_ bytes available in `read` than we've read. Do we want to try reading one more byte, and see that we get EOF?

The current code is safe - we check the digest, so we won't write incorrect data for a particular digest, but it may miss extra appended bytes.

Reply:
Yea, great point. I thought about this, and I figured that a file lengthening wasn't worth checking/erroring for, since the stored data was still valid. I'll add a comment here to that effect.

In terms of the graph, invalidation will cause us to come and recheck the file later.